src/msg/producer/ref_counted.go

```diff
@@ -31,24 +31,24 @@ type OnFinalizeFn func(rm *RefCountedMessage)
 
 // RefCountedMessage is a reference counted message.
 type RefCountedMessage struct {
-	sync.RWMutex
+	mu sync.RWMutex
 	Message
 
 	size         uint64
 	onFinalizeFn OnFinalizeFn
 
-	refCount            *atomic.Int32
-	isDroppedOrConsumed *atomic.Bool
+	// RefCountedMessage must not be copied by value due to RWMutex,
+	// safe to store values here and not just pointers
+	refCount            atomic.Int32
+	isDroppedOrConsumed atomic.Bool
 }
 
 // NewRefCountedMessage creates RefCountedMessage.
 func NewRefCountedMessage(m Message, fn OnFinalizeFn) *RefCountedMessage {
 	return &RefCountedMessage{
-		Message:             m,
-		refCount:            atomic.NewInt32(0),
-		size:                uint64(m.Size()),
-		onFinalizeFn:        fn,
-		isDroppedOrConsumed: atomic.NewBool(false),
```

Comment:
keep this?

Reply:
Sorry, I rebased, so lost context - by changing underlying type from `*atomic.XXX` to just `atomic.XXX` there's no longer a need to initialize those - as long as the `RefCountedMessage` and `message` will not be copied by value (and they aren't), it's perfectly safe to use atomics by value (they use a pointer to underlying value in `atomic.XXX` struct).

Reply:
unless the struct embedding atomic value wrapper is going to be passed around by value, there's no need to allocate the value itself on the heap somewhere.

Reply:
gotcha